File: src/fani/report.py
"""Run reports.

Two artefacts per run: ``report.json`` for machines (the scheduler, a later
gate, a dashboard) and ``report.md`` for the human who has to decide what to do
about a ``needs_human`` result. Both are written even when the run fails, so a
failed run is never a silent one.
"""

import json
import platform
import time
from pathlib import Path

from .orchestrator import EXIT_CODES, LangOutcome, Status

_ORDER = [Status.ERROR, Status.NEEDS_HUMAN, Status.PARTIAL, Status.OK]

_LABEL = {
    Status.OK: "ok",
    Status.NEEDS_HUMAN: "needs human",
    Status.PARTIAL: "partial",
    Status.ERROR: "error",
}


def overall(outcomes: list[LangOutcome]) -> Status:
    """The worst status wins; an empty run is ok."""
    for status in _ORDER:
        if any(o.status is status for o in outcomes):
            return status
    return Status.OK


def exit_code(outcomes: list[LangOutcome]) -> int:
    return EXIT_CODES[overall(outcomes)]


def build(outcomes: list[LangOutcome], started_at: float, config_path: Path) -> dict:
    dispatch = [d for o in outcomes for d in o.dispatch]
    failed = [d for d in dispatch if not d.ok]
    return {
        "schema": 1,
        "config": str(config_path),
        "started_at": time.strftime("%Y-%m-%dT%H:%M:%S%z", time.localtime(started_at)),
        "duration_s": round(time.time() - started_at, 3),
        "host": platform.node(),
        "status": overall(outcomes).value,
        "exit_code": exit_code(outcomes),
        "totals": {
            "languages": len(outcomes),
            "files_written": sum(len(o.written) for o in outcomes),
            "conflicts": sum(len(o.conflicts) for o in outcomes),
            "findings": sum(len(o.findings) for o in outcomes),
            "agent_calls": len(dispatch),
            "agent_failures": len(failed),
            "repair_rounds": sum(o.repair_rounds for o in outcomes),
            "remaining_tasks": sum(o.remaining_tasks for o in outcomes),
            "fuzzy_matched": sum(o.fuzzy_matched for o in outcomes),
            "commits": sum(1 for o in outcomes if o.published.commit),
        },
        "languages": [o.to_dict() for o in outcomes],
    }


def render(data: dict) -> str:
    t = data["totals"]
    lines = [
        f"# fani run — {data['status']}",
        "",
        f"- started: {data['started_at']} on {data['host']}",
        f"- duration: {data['duration_s']:.1f}s",
        f"- config: `{data['config']}`",
        f"- files written: {t['files_written']} · conflicts: {t['conflicts']} · "
        f"findings: {t['findings']}",
        f"- agent calls: {t['agent_calls']} ({t['agent_failures']} failed) · "
        f"repair rounds: {t['repair_rounds']} · reused from memory: {t['fuzzy_matched']}",
        "",
    ]

    if data["languages"]:
        lines += [
            "| repo | lang | status | written | findings | agent calls | commit | time |",
            "| --- | --- | --- | --- | --- | --- | --- | --- |",
        ]
        for lang in data["languages"]:
            label = _LABEL[Status(lang["status"])]
            pub = lang.get("published", {})
            commit = f"`{pub['commit'][:9]}` on `{pub['branch']}`" if pub.get("commit") else "—"
            lines.append(
                f"| {Path(lang['repo']).name} | {lang['lang']} | {label} | "
                f"{len(lang['written'])} | {len(lang['findings'])} | "
                f"{len(lang['dispatch'])} | {commit} | {lang['duration_s']:.1f}s |"
            )
        lines.append("")

    for lang in data["languages"]:
        if Status(lang["status"]) is Status.OK and not lang["findings"]:
            continue
        lines += [f"## {Path(lang['repo']).name} — {lang['lang']}", "", lang["message"] or "", ""]
        if lang["conflicts"]:
            lines.append("Hand-edited translations (not overwritten):")
            lines += [f"- `{c.get('path', c)}`" for c in lang["conflicts"]]
            lines.append("")
        if lang["findings"]:
            lines.append("Findings:")
            for f in lang["findings"][:40]:
                if isinstance(f, dict):
                    where = f.get("file") or f.get("path") or "?"
                    lines.append(
                        f"- `{where}` {f.get('severity', '')} {f.get('code', '')}: "
                        f"{f.get('message', '')}".rstrip()
                    )
                else:
                    lines.append(f"- {f}")
            if len(lang["findings"]) > 40:
                lines.append(f"- … {len(lang['findings']) - 40} more (see report.json)")
            lines.append("")
        failed = [d for d in lang["dispatch"] if not d["ok"]]
        if failed:
            lines.append("Failed agent calls:")
            lines += [
                f"- `{d['task_id']}` {d['code']}: {d['message']}" for d in failed[:20]
            ]
            lines.append("")

    return "\n".join(lines).rstrip() + "\n"


def write(outcomes: list[LangOutcome], out_dir: Path, started_at: float,
          config_path: Path) -> dict:
    data = build(outcomes, started_at, config_path)
    out_dir.mkdir(parents=True, exist_ok=True)
    (out_dir / "report.json").write_text(
        json.dumps(data, ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
    )
    (out_dir / "report.md").write_text(render(data), encoding="utf-8")
    return data

File: src/report.rs
use crate::model::{LangOutcome, Status};
use anyhow::Result;
use chrono::{DateTime, Local};
use serde_json::{Value, json};
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};

pub fn overall(outcomes: &[LangOutcome]) -> Status {
    for status in [
        Status::Error,
        Status::NeedsHuman,
        Status::Partial,
        Status::Ok,
    ] {
        if outcomes.iter().any(|o| o.status == status) {
            return status;
        }
    }
    Status::Ok
}

pub fn build(
    outcomes: &[LangOutcome],
    started: DateTime<Local>,
    config: &Path,
    database_paths: &[PathBuf],
) -> Value {
    let dispatch: Vec<_> = outcomes.iter().flat_map(|o| o.dispatch.iter()).collect();
    json!({
        "schema": 2,
        "config": config.display().to_string(),
        "databases": database_paths.iter().map(|p| p.display().to_string()).collect::<Vec<_>>(),
        "started_at": started.to_rfc3339(),
        "duration_s": (Local::now() - started).num_milliseconds() as f64 / 1000.0,
        "host": std::env::var("HOSTNAME").unwrap_or_else(|_| "unknown".into()),
        "status": overall(outcomes).as_str(),
        "exit_code": overall(outcomes).exit_code(),
        "totals": {
            "languages": outcomes.len(),
            "files_written": outcomes.iter().map(|o| o.written.len()).sum::<usize>(),
            "conflicts": outcomes.iter().map(|o| o.conflicts.len()).sum::<usize>(),
            "findings": outcomes.iter().map(|o| o.findings.len()).sum::<usize>(),
            "agent_calls": dispatch.len(),
            "task_dispatches": dispatch.len(),
            "agent_attempts": dispatch.iter().map(|d| d.attempts).sum::<usize>(),
            "agent_retries": dispatch.iter().map(|d| d.attempts.saturating_sub(1)).sum::<usize>(),
            "agent_failures": dispatch.iter().filter(|d| !d.ok).count(),
            "repair_rounds": outcomes.iter().map(|o| o.repair_rounds).sum::<usize>(),
            "remaining_tasks": outcomes.iter().map(|o| o.remaining_tasks).sum::<usize>(),
            "fuzzy_matched": outcomes.iter().map(|o| o.fuzzy_matched).sum::<usize>(),
            "commits": outcomes.iter().filter(|o| !o.published.commit.is_empty()).count(),
        },
        "languages": outcomes,
    })
}

pub fn render(data: &Value) -> String {
    let totals = &data["totals"];
    let mut lines = vec![
        format!(
            "# fani run — {}",
            data["status"].as_str().unwrap_or("error")
        ),
        String::new(),
        format!(
            "- started: {} on {}",
            data["started_at"].as_str().unwrap_or(""),
            data["host"].as_str().unwrap_or("")
        ),
        format!(
            "- duration: {:.1}s",
            data["duration_s"].as_f64().unwrap_or(0.0)
        ),
        format!("- config: `{}`", data["config"].as_str().unwrap_or("")),
        format!(
            "- files written: {} · conflicts: {} · findings: {}",
            totals["files_written"], totals["conflicts"], totals["findings"]
        ),
        format!(
            "- task dispatches: {} · agent attempts: {} ({} failed tasks, {} retries) · repair rounds: {} · reused from memory: {}",
            totals["task_dispatches"],
            totals["agent_attempts"],
            totals["agent_failures"],
            totals["agent_retries"],
            totals["repair_rounds"],
            totals["fuzzy_matched"]
        ),
        String::new(),
    ];
    let languages = data["languages"].as_array().cloned().unwrap_or_default();
    if !languages.is_empty() {
        lines.extend([
            "| repo | lang | status | written | findings | agent calls | commit | time |".into(),
            "| --- | --- | --- | --- | --- | --- | --- | --- |".into(),
        ]);
        for lang in &languages {
            let repo = Path::new(lang["repo"].as_str().unwrap_or("?"))
                .file_name()
                .and_then(|x| x.to_str())
                .unwrap_or("?");
            let commit = lang["published"]["commit"]
                .as_str()
                .filter(|x| !x.is_empty())
                .map(|sha| {
                    format!(
                        "`{}` on `{}`",
                        &sha[..sha.len().min(9)],
                        lang["published"]["branch"].as_str().unwrap_or("")
                    )
                })
                .unwrap_or_else(|| "—".into());
            lines.push(format!(
                "| {repo} | {} | {} | {} | {} | {} | {commit} | {:.1}s |",
                lang["lang"].as_str().unwrap_or(""),
                lang["status"].as_str().unwrap_or(""),
                lang["written"].as_array().map_or(0, Vec::len),
                lang["findings"].as_array().map_or(0, Vec::len),
                lang["dispatch"].as_array().map_or(0, Vec::len),
                lang["duration_s"].as_f64().unwrap_or(0.0)
            ));
        }
        lines.push(String::new());
    }
    for lang in &languages {
        let has_failed_calls = lang["dispatch"]
            .as_array()
            .is_some_and(|calls| calls.iter().any(|call| call["ok"] == false));
        let clean = lang["status"] == "ok"
            && lang["findings"].as_array().is_none_or(Vec::is_empty)
            && !has_failed_calls;
        if clean {
            continue;
        }
        let repo = Path::new(lang["repo"].as_str().unwrap_or("?"))
            .file_name()
            .and_then(|x| x.to_str())
            .unwrap_or("?");
        lines.extend([
            format!("## {repo} — {}", lang["lang"].as_str().unwrap_or("")),
            String::new(),
            lang["message"].as_str().unwrap_or("").into(),
            String::new(),
        ]);
        if let Some(conflicts) = lang["conflicts"].as_array().filter(|x| !x.is_empty()) {
            lines.push("Hand-edited translations (not overwritten):".into());
            for c in conflicts {
                lines.push(format!(
                    "- `{}`",
                    c.get("path").and_then(Value::as_str).unwrap_or("?")
                ));
            }
            lines.push(String::new());
        }
        let failed_calls: Vec<_> = lang["dispatch"]
            .as_array()
            .into_iter()
            .flatten()
            .filter(|call| call["ok"] == false)
            .collect();
        if !failed_calls.is_empty() {
            lines.push("Failed agent calls:".into());
            for call in failed_calls {
                lines.push(format!(
                    "- `{}` [{}]: {}",
                    call["task_id"].as_str().unwrap_or("?"),
                    call["code"].as_str().unwrap_or("DSP-EXIT"),
                    call["message"].as_str().unwrap_or("")
                ));
            }
            lines.push(String::new());
        }
        if let Some(findings) = lang["findings"].as_array().filter(|x| !x.is_empty()) {
            lines.push("Findings:".into());
            for f in findings.iter().take(40) {
                lines.push(format!(
                    "- `{}` {} {}: {}",
                    f.get("file")
                        .or_else(|| f.get("path"))
                        .and_then(Value::as_str)
                        .unwrap_or("?"),
                    f["severity"].as_str().unwrap_or(""),
                    f["code"].as_str().unwrap_or(""),
                    f["message"].as_str().unwrap_or("")
                ));
            }
            if findings.len() > 40 {
                lines.push(format!(
                    "- … {} more (see report.json)",
                    findings.len() - 40
                ));
            }
            lines.push(String::new());
        }
    }
    lines.join("\n").trim_end().to_string() + "\n"
}

fn atomic_write(path: &Path, content: &[u8]) -> Result<()> {
    let parent = path.parent().unwrap_or_else(|| Path::new("."));
    let mut temp = tempfile::NamedTempFile::new_in(parent)?;
    temp.write_all(content)?;
    temp.as_file().sync_all()?;
    temp.persist(path).map_err(|e| e.error)?;
    Ok(())
}

pub fn write(
    outcomes: &[LangOutcome],
    dir: &Path,
    started: DateTime<Local>,
    config: &Path,
    database_paths: &[PathBuf],
) -> Result<Value> {
    let data = build(outcomes, started, config, database_paths);
    fs::create_dir_all(dir)?;
    atomic_write(
        &dir.join("report.json"),
        (serde_json::to_string_pretty(&data)? + "\n").as_bytes(),
    )?;
    atomic_write(&dir.join("report.md"), render(&data).as_bytes())?;
    Ok(data)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::model::LangOutcome;
    use tempfile::tempdir;

    #[test]
    fn worst_status_and_reports_match_contract() {
        let mut ok = LangOutcome::new(Path::new("/repo/docs"), "zh-CN");
        ok.message = "done".into();
        let mut partial = LangOutcome::new(Path::new("/repo/docs"), "ja");
        partial.status = Status::Partial;
        partial.remaining_tasks = 2;
        assert_eq!(overall(&[ok.clone(), partial.clone()]).exit_code(), 3);
        let tmp = tempdir().unwrap();
        let data = write(
            &[ok, partial],
            tmp.path(),
            Local::now(),
            Path::new("fani.toml"),
            &[],
        )
        .unwrap();
        assert_eq!(data["status"], "partial");
        assert!(tmp.path().join("report.json").is_file());
        assert!(
            fs::read_to_string(tmp.path().join("report.md"))
                .unwrap()
                .starts_with("# fani run")
        );
    }
}
